cql3/statements/select_statement.cc

```diff
@@ -1499,70 +1502,132 @@ void select_statement::verify_ordering_is_allowed(const restrictions::statement_
+        if (column.type->is_reversed()) {
+            return true;
+        } else {
+            return false;
+        }
+    } else { // descending ordering
+        if (column.type->is_reversed()) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+}
+
+bool select_statement::is_ordering_reversed(const prepared_orderings_type& orderings) const {
+    if (orderings.empty()) {
+        return false;
+    }
+
+    return are_column_select_results_reversed(*orderings[0].first, orderings[0].second);
+}
+
+void select_statement::verify_ordering_is_valid(const prepared_orderings_type& orderings,
+                                                const schema& schema,
+                                                const restrictions::statement_restrictions& restrictions) const {
+    if (orderings.empty()) {
         return;
     }
 
-    for (auto&& def : schema.partition_key_columns()) {
-        if (!selection.has_column(def)) {
-            throw exceptions::invalid_request_exception(format("SELECT DISTINCT queries must request all the partition key columns (missing {})", def.name_as_text()));
+    // Verify that columns are in the same order as in schema definition
+    for (size_t i = 0; i + 1 < orderings.size(); i++) {
+        if (orderings[i].first->component_index() >= orderings[i + 1].first->component_index()) {
+            throw exceptions::invalid_request_exception("Order by currently only supports the ordering of columns following their declared order in the PRIMARY KEY");
         }
     }
-}
 
-void select_statement::handle_unrecognized_ordering_column(const column_identifier& column) const
-{
-    if (contains_alias(column)) {
-        throw exceptions::invalid_request_exception(format("Aliases are not allowed in order by clause ('{}')", column));
+    // We only allow leaving select results unchanged or reversing them. Find out if they should be reversed.
+    bool is_reversed = is_ordering_reversed(orderings);
+
+    // Now verify that
+    // 1. Each column in the specified clustering prefix either has an ordering, or an EQ restriction.
+    //    When there is an EQ restriction ordering is not needed because only a single value is possible.
+    // 2. The ordering leaves the selected rows unchanged or reverses the normal result.
+    //    For that to happen the ordering must reverse all columns or none at all.
+    uint32_t max_ck_index = orderings.back().first->component_index();
+    auto orderings_iterator = orderings.begin();
+    for (uint32_t ck_column_index = 0; ck_column_index <= max_ck_index; ck_column_index++) {
+        if (orderings_iterator->first->component_index() != ck_column_index) {
+            // We don't have ordering for this column. Fail with an error.
+            throw exceptions::invalid_request_exception("Order by currently only supports the ordering of columns following their declared order in the PRIMARY KEY");
```

Comment:
Is this the right error message to print here? Especially if (I think you said) you later want to allow playing with the order, then the issue here isn't that the order is wrong but that one of the components is missing.

Reply:
I kept this message, because it was the one expected in the tests.
But you are right, it makes no sense.
Changed it to something more sensible and removed checking error contents from the test.

Reply:
Did you push the changed version? I don't seem to see a change here? (but I do see that the tests changed)

Reply:
I changed it in a later commit, but forgot to do it in the previous one. Fixed.